pkg/dispers/conceptindexor.go

```diff
@@ -0,0 +1,198 @@
+type ConceptDoc struct {
+	ConceptID  string `json:"_id,omitempty"`
+	ConceptRev string `json:"_rev,omitempty"`
+	Concept    string `json:"concept,omitempty"`
+	Hash       string `json:"hash,omitempty"`
+}
+
+// ID is used to get ID
+func (t *ConceptDoc) ID() string {
+	return t.ConceptID
+}
+
+// Rev is used to get Rev
+func (t *ConceptDoc) Rev() string {
+	return t.ConceptRev
+}
+
+// DocType is used to get DocType
+func (t *ConceptDoc) DocType() string {
+	return doctypeSalt
+}
+
+// Clone is used to create another ConceptDoc from this ConceptDoc
+func (t *ConceptDoc) Clone() couchdb.Doc {
+	cloned := *t
+	return &cloned
+}
+
+// SetID is used to set the Doc ID
+func (t *ConceptDoc) SetID(id string) {
+	t.ConceptID = id
+}
+
+// SetRev is used to set Rev
+func (t *ConceptDoc) SetRev(rev string) {
+	t.ConceptRev = rev
+}
+
+func saveConcept(concept string) error {
+
+	hash, err := createHash(concept)
+	if err != nil {
+		return err
+	}
+
+	conceptDoc := &ConceptDoc{
+		Concept: concept,
+		Hash:    hash,
+	}
+	return couchdb.CreateDoc(prefixerCI, conceptDoc)
+}
+
+func getHash(concept string) (string, error) {
+
+	// Create index to run a mango query
+	hash := ""
+
+	// Precise and run the mango query
+	var out []ConceptDoc
+	req := &couchdb.FindRequest{Selector: mango.Equal("concept", concept)}
+	err := couchdb.FindDocs(prefixerCI, doctypeSalt, req, &out)
+	if err != nil {
+		return hash, err
+	}
+
+	// Create a salt if no salt in the database for this concept
+	if len(out) == 1 {
+		hash = out[0].Hash
+	} else if len(out) == 0 {
+		return "", errors.New("Concept Indexor : no existing salt for this concept")
+	} else {
+		return "", errors.New("Concept Indexor : several salts for this concept")
+	}
+
+	return hash, err
+}
+
+func createHash(str string) (string, error) {
+
+	salt := string(crypto.GenerateRandomBytes(defaultLenSalt))
+
+	h := sha256.New()
+	h.Write([]byte(str + salt))
+	hash := h.Sum([]byte("concept"))
+
+	return string(hash), nil
+}
+
+func isConceptExisting(concept string) (bool, error) {
+
+	// Precise and run the mango query
+	var out []ConceptDoc
+	req := &couchdb.FindRequest{Selector: mango.Equal("concept", concept)}
+	err := couchdb.FindDocs(prefixerCI, doctypeSalt, req, &out)
+	if err != nil {
+		return false, err
+	}
+
+	if len(out) > 0 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// CreateConcept checks if concept exists in db. If yes, return error. If no, create the salt, save the concept in db and return the hash.
+func CreateConcept(in *dispers.Concept) error {
+
+	if err := couchdb.EnsureDBExist(prefixerCI, doctypeSalt); err != nil {
+		return err
+	}
+
+	// Get salt with concept
+	isExisting, err := isConceptExisting(in.Concept)
+	if err != nil {
+		return err
+	}
+
+	if isExisting {
+		return errors.New("Concept is already existing")
+	}
+	if err = saveConcept(in.Concept); err != nil {
+		return err
+	}
+	in.Hash, err = getHash(in.Concept)
```

Comment:
Instead of trying the find the hash in database just after created it, it would be more efficient to return the created doc from `saveConcept` and get the created hash from it